### ISSUE TITLE
vtgate should have liveness _and_ readiness probes (is using with mysql protocol access)

### ISSUE THREAD

**Issue by sjmudd** [vitessio/vitess#4072]
Please search the existing issues for relevant feature requests, and use the [reaction feature](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to add upvotes to pre-existing requests.

#### Feature Description

Kubernetes suggests that processes should have a liveness and readiness end point. see: https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-probes/.
One of these is to check the process is up and running, the other that it is ready to service requests.

vtgate if used with the mysql protocol doesn't provide a way for a "health tester" to know if the process should not take traffic. A reason for this would be because you want to "drain connections" so you can do maintenance.  Right now I'm working around this by handling it outside of vtgate but having a standard way to do this would be good.

Thoughts on how to accomplish this might be:
* provide an http endpoint to see if the proces is ready to accept requests, and an endpoint to set this state (enable/disable). [ By default vtgate would be expected to accept requests. ]  Follow whatever might be the standard endpoint naming used with kubernetes checks.

With https://github.com/vitessio/vitess/pull/4070 we will be able to monitor users connected to Vitess by the MySQL protocol and can wait for the number to drain to 0 if needed.  Adding this functionality would mean the process of swapping out a vtgate would be cleaner.

If this can be handled better another way then please point me at suggestions on how to do that.

**Comment by derekperkins** [comment]
That would be great. This is what we're currently using for the Kubernetes checks.
https://github.com/vitessio/vitess/blob/master/helm/vitess/templates/_vtgate.tpl#L76-L87

**Comment by sjmudd** [comment]
i also checked the code. there is a readiness check.
https://github.com/vitessio/vitess/blob/master/go/vt/vtgate/vtgate.go#L276
This is always healthy, so all that's needed is to provide a away to make the readiness check fail.

That seems to indicate to me that we need some sort of http api call to set the two values "ok" or "not ok", I'm not sure which would be the best way of doing that or which api end point.

Maybe it would be possible to write to the `/debug/health` end point?
I see there's a `acl.MONITORING` check so it probably should be safe to allow writes to that path.

**Comment by derekperkins** [comment]
That function is probably good as is for the liveness check.

For the readiness check, what situations would you define as making it unready? Are you anticipating that you'll hit the PUT endpoint externally to Vitess?

**Comment by sjmudd** [comment]
Yes. My intention prior to say upgrading `vtgate` would be to do a PUT to `/debug/health` with "not ok" and for it to "stick". I'll be able to monitor connected users by username with Scott's patch: https://github.com/vitessio/vitess/pull/4070 and when it's low enough I should be able to safely stop vtgate as my internal load balancing infrastructure on seeing `/debug/health` as "not ok" will stop forwarding requests to the `vtgate` in question.

**Comment by sjmudd** [comment]
but maybe this change will mean that code only checking if vtgate is alive and looking for "ok" will be confused if it says "not ok".  Does that mean I nead to provide a different end point? Not quite sure what's best.
